From 8313885b3c68c56695ea7e27f5c7652bcb9b096e Mon Sep 17 00:00:00 2001
From: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Date: Mon, 23 Jan 2023 14:28:24 +0530
Subject: [PATCH] cfg80211: Allow non public action frames to be transmitted
 with link BSS

In case of an MLD if User-space has provided a link_id to action frame
tx request, allow transmission of the frame in that link, Currently
action frames only with ML address as A3 are allowed, but tx for a
non-ml STA on any link of an ML BSS can happen with link address
as A3.

Signed-off-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
---
 net/wireless/mlme.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/net/wireless/mlme.c b/net/wireless/mlme.c
index 42576e8..5ad4a19 100644
--- a/net/wireless/mlme.c
+++ b/net/wireless/mlme.c
@@ -739,7 +739,10 @@ int cfg80211_mlme_mgmt_tx(struct cfg80211_registered_device *rdev,
 		case NL80211_IFTYPE_P2P_GO:
 		case NL80211_IFTYPE_AP_VLAN:
 			if (!ether_addr_equal(mgmt->bssid, wdev_address(wdev)))
-				err = -EINVAL;
+				if (params->link_id < 0 ||
+				    !ether_addr_equal(mgmt->bssid,
+						      wdev->links[params->link_id].addr))
+					err = -EINVAL;
 			break;
 		case NL80211_IFTYPE_MESH_POINT:
 			if (!ether_addr_equal(mgmt->sa, mgmt->bssid)) {
-- 
2.17.1

